Parallel, multi-GPU pairwise inference

schafferde opened the pull request.

schafferde commented on dscript/commands/predict_par.py:
```diff
@@ -0,0 +1,228 @@
+"""
+Make new predictions with a pre-trained model. One of --seqs or --embeddings is required.
```
Yes, deleted

schafferde commented on dscript/commands/par_worker.py:
```diff
@@ -0,0 +1,98 @@
+from ..models.interaction import DSCRIPTModel
+
+#Worker process function for parallel prediction
+#Because one can't pass an open file, logFile will always be none.
```
Done

schafferde commented on dscript/commands/predict.py:
```diff
@@ -66,6 +67,12 @@ def add_args(parser):
+    parser.add_argument(
+        "--load_proc",
+        type=int,
+        default=-1,
```
Yes, changed to 16 for new modes and 32 for legacy mode

schafferde commented on dscript/commands/par_worker.py:
```diff
@@ -0,0 +1,98 @@
+        except FileNotFoundError:
+            log(f"Model {modelPath} not found", file=logFile, print_also=True)
+            #logFile.close()
+            sys.exit(6) #Is it bad to call this from multiple processes?
```
I moved a file-exists check to main, but loading does need to be done individually because a) the model goes directly to each GPU as is and b) it would be costly to pass a whole model through the process initialization. 
6 is an arbitrary positive integer

schafferde commented on dscript/commands/predict_bipartite.py:
```diff
@@ -0,0 +1,338 @@
+        try:
+            log(f"Loading protein IDs from {csvPath}", file=logFile, print_also=True)
+            with open(csvPath) as f:
+                self.all_prots = [line.strip() for line in f if line and not line.isspace()]
```
Same response

samsledje commented:
@schafferde I'm trying to run this locally on the test cases and it is just hanging, even with a very small number of pairs-- you can see what's failing in the check above. Any thoughts?

samsledje commented:
Okay, this seems to be linked to model loading and `weights_only=False` -- I can fix this, but it raises the concern that if a process throws an error, this is not reported and the process appears to hang. Is there any way to check whether an error has popped up in a GPU computing process and catch this/report back to the user? I think this silent failure would occur if, for example, the user has a typo in their local file path.

schafferde commented:
@samsledje Could you provide a bit more information about what you observed? Looking at the build tests, the error I see seems to be that the call is for the old `predict` signature (now `predict_serial`).

I agree that currently certain errors in one process could cause a hang. I think the simplest solution is to try to validate input data and catch errors during execution (which may be incomplete). For example, certain errors during prediction are caught and result in an output of -1 for that pair. For local file paths, these should be validated in the main process before the GPU processes are started. (For the model path, I moved this check from the GPU processes per a previous comment.)

samsledje commented:
> @samsledje Could you provide a bit more information about what you observed? Looking at the build tests, the error I see seems to be that the call is for the old `predict` signature (now `predict_serial`).
> 
> I agree that currently certain errors in one process could cause a hang. I think the simplest solution is to try to validate input data and catch errors during execution (which may be incomplete). For example, any error during the actual prediction step is caught and results in an output of -1 for that pair. For local file paths, these should be validated in the main process before the GPU processes are started. (For the model path, I moved this check from the GPU processes per a previous comment.)

I guess this wouldn't be an issue for the user provided we provide bug-free code-- but it makes debugging a lot harder that these processes seem to fail silently.

samsledje commented:
I think this should fix it-- it runs for me locally-- but please test on your end?

schafferde commented:
I want to make a few tweaks, improve validation, and run tests. One sec.

schafferde commented:
Ok, I'm done, sorry for the delay. I made some tweaks to improve validation and simplify. I also moved the device parsing to `utils.py` and added it to `predict_bipartite` as well.
Finally, I had to update my pytorch to 1.13 and I tried to update requirements accordingly.
